### PR TITLE
Switch pretty printer to block-based indentation

### PR DESCRIPTION
**Pull request by dtolnay**
This PR backports https://github.com/dtolnay/prettyplease/commit/401d60c04213e6c66565e0e69a95b4588db5fdba from the `prettyplease` crate into `rustc_ast_pretty`.

A before and after:

```diff
- let res =
-     ((::alloc::fmt::format as
-          for<'r> fn(Arguments<'r>) -> String {format})(((::core::fmt::Arguments::new_v1
-                                                             as
-                                                             fn(&[&'static str], &[ArgumentV1]) -> Arguments {Arguments::new_v1})((&([("test"
-                                                                                                                                          as
-                                                                                                                                          &str)]
-                                                                                                                                        as
-                                                                                                                                        [&str; 1])
-                                                                                                                                      as
-                                                                                                                                      &[&str; 1]),
-                                                                                                                                  (&([]
-                                                                                                                                        as
-                                                                                                                                        [ArgumentV1; 0])
-                                                                                                                                      as
-                                                                                                                                      &[ArgumentV1; 0]))
-                                                            as
-                                                            Arguments))
-         as String);
+ let res =
+     ((::alloc::fmt::format as
+             for<'r> fn(Arguments<'r>) -> String {format})(((::core::fmt::Arguments::new_v1
+                 as
+                 fn(&[&'static str], &[ArgumentV1]) -> Arguments {Arguments::new_v1})((&([("test"
+                             as &str)] as [&str; 1]) as
+                 &[&str; 1]),
+             (&([] as [ArgumentV1; 0]) as &[ArgumentV1; 0])) as
+             Arguments)) as String);
```

Previously the pretty printer would compute indentation always relative to whatever column a block begins at, like this:

```rust
fn demo(arg1: usize,
        arg2: usize);
```

This is never the thing to do in the dominant contemporary Rust style. Rustfmt's default and the style used by the vast majority of Rust codebases is block indentation:

```rust
fn demo(
    arg1: usize,
    arg2: usize,
);
```

where every indentation level is a multiple of 4 spaces and each level is indented relative to the indentation of the previous line, not the position that the block starts in.

By itself this PR doesn't get perfect formatting in all cases, but it is the smallest possible step in clearly the right direction. More backports from `prettyplease` to tune the ibox/cbox indent levels around various AST node types are upcoming.

### CODE REVIEW

**Review by nagisa — APPROVED** [review]
Some of examples do indeed look much more conventional, nice.

I don't actually know enough about ast_pretty (and I bet the sentiment will be shared across T-compiler in general) to have much idea what some of the minute changes to the code affect, but I'm quite comfortable r+ing this regardless given that all these changes can affect is whitespace.

My question however more broadly is – is there a reason why we're looking to backport the changes from prettyplease to ast_pretty vs just migrating rustc to use prettyplease in the first place?

**Comment by bjorn3** [comment]
> My question however more broadly is – is there a reason why we're looking to backport the changes from prettyplease to ast_pretty vs just migrating rustc to use prettyplease in the first place?

Prettyplease uses syn and thus proc_macro `TokenStream`, while rustc directly uses the pretty printer on the AST and HIR.

File: compiler/rustc_ast_pretty/src/pp.rs [hunk 1]
```diff
@@ -146,6 +146,22 @@ pub enum Breaks {
     Inconsistent,
 }
 
+#[derive(Clone, Copy)]
+enum IndentStyle {
+    /// Vertically aligned under whatever column this block begins at.
+    ///
+    ///     fn demo(arg1: usize,
+    ///             arg2: usize);
+    Visual,
```
[the remark this answers is not in the document]
Comment (dtolnay) [reply compiler/rustc_ast_pretty/src/pp.rs:155]: This is only used by trailing block comments, which align visually under the `/*`. I would be very happy to delete that test and revert 269b5a454ff7d33ea9ea7bdf577e94c7de6ae1bc.

https://github.com/rust-lang/rust/blob/ecf72996eda4f8af19b0ca7235c6f62e0245a313/src/test/pretty/block-comment-trailing-whitespace2.rs#L4-L5

**Comment by dtolnay** [comment]
Yeah there's no way that `prettyplease` would be useful for printing a rustc AST.

**Comment by dtolnay** [comment]
To clarify from https://github.com/rust-lang/rust/pull/93155#discussion_r790136025: @nagisa would you prefer if 269b5a454ff7d33ea9ea7bdf577e94c7de6ae1bc were removed from this PR? I don't expect it to be a large maintenance cost based on the other pretty printer changes I've written after this one and the small rate of change in the pretty printer in general, but it's also not a large benefit because block comments are pretty rare in modern Rust, so it's tough to gauge the right tradeoff.

**Comment by nagisa** [comment]
Nah, I'm fine with it being in, personally. I guess I forgot to give r=me after reviewing, despite approving the PR…

@bors r+

**Comment by dtolnay** [comment]
- Rebased to resolve conflict with #90891 in `src/test/ui/attributes/key-value-expansion.stderr`

**Pull request merged**